web3/encoding.nim

```diff
@@ -9,13 +9,228 @@
 
 import
   std/macros,
-  stint, ./eth_api_types, stew/[assign2, byteutils]
+  stint, ./eth_api_types, stew/[assign2, byteutils, endians2]
 
+type
+  AbiEncoder* = object
+    stack: seq[Tuple]
+  Tuple = object
+    bytes: seq[byte]
+    postponed: seq[Split]
```

Comment:
something to consider is whether https://github.com/status-im/nim-faststreams?tab=readme-ov-file#delayed-writes can be used here - ie make AbiEncoder write to a memoryOutput and use it to manage encoding - this can potentially reduce the amount of logic in web3 and we're already using faststreams elsewhere for similar purposes